packages/design/src/components/popup/RectPopup.tsx

```diff
@@ -0,0 +1,140 @@
+/**
+ * Copyright 2023-present DreamNum Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useEffect, useRef, useState } from 'react';
+import styles from './index.module.less';
+
+interface IAbsolutePosition {
+    left: number;
+    right: number;
+    top: number;
+    bottom: number;
+}
+
+export interface IRectPopupProps {
+    children?: React.ReactNode;
+
+    /**
+     * the anchor element bounding rect
+     */
+    anchorRect: IAbsolutePosition;
+
+    direction?: 'horizontal' | 'vertical';
+
+    mask?: boolean;
+
+    onMaskClick?: React.MouseEventHandler<HTMLDivElement>;
+
+    onClickOther?: (e: MouseEvent) => void;
+}
+
+const calcHorizontalPopupPosition = (position: IAbsolutePosition, width: number, height: number, containerWidth: number, containerHeight: number): Partial<IAbsolutePosition> => {
```

Comment:
These functions seems to be reusable in other components. We can extract them to a util dir.

Reply:
reusing these logic by using `RectPopup.calcPopupPosition`.